src/main/java/org/skriptlang/skript/registration/DefaultSyntaxInfosImpl.java

```diff
@@ -0,0 +1,195 @@
+ *
+ *  Skript is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package org.skriptlang.skript.registration;
+
+import ch.njol.skript.lang.ExpressionType;
+import com.google.common.base.MoreObjects;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.Nullable;
+import org.skriptlang.skript.lang.Priority;
+import org.skriptlang.skript.lang.entry.EntryValidator;
+
+import java.util.Collection;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+final class DefaultSyntaxInfosImpl {
+
+	public static class ExpressionImpl<E extends ch.njol.skript.lang.Expression<R>, R>
+		extends SyntaxInfoImpl<E> implements DefaultSyntaxInfos.Expression<E, R> {
+
+		private final Class<R> returnType;
+		private final ExpressionType expressionType;
+
+		ExpressionImpl(
+			SyntaxOrigin origin, Class<E> type, @Nullable Supplier<E> supplier,
+			Collection<String> patterns, Class<R> returnType, ExpressionType expressionType
+		) {
+			super(origin, type, supplier, patterns);
+			if (returnType.isAnnotation() || returnType.isArray() || returnType.isPrimitive()) {
+				throw new IllegalArgumentException("returnType must be a normal type");
+			}
+			this.returnType = returnType;
+			this.expressionType = expressionType;
+		}
+
+		@Override
+		public Class<R> returnType() {
+			return returnType;
+		}
+
+		@Override
+		public ExpressionType expressionType() {
+			return expressionType;
+		}
+
+		@Override
+		public boolean equals(Object other) {
+			if (!(other instanceof Expression) || !super.equals(other)) {
+				return false;
+			}
+			ExpressionImpl<?, ?> expression = (ExpressionImpl<?, ?>) other;
+			return returnType() == expression.returnType() &&
+					Objects.equals(expressionType(), expression.expressionType());
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hash(origin(), type(), patterns(), returnType(), expressionType());
+		}
+
+		@Override
+		public String toString() {
+			return MoreObjects.toStringHelper(this)
+					.add("origin", origin())
+					.add("type", type())
+					.add("patterns", patterns())
+					.add("returnType", returnType())
+					.add("expressionType", expressionType())
+					.toString();
+		}
+
+		@Override
+		public Priority priority() {
+			return new Priority(super.priority().getPriority() | expressionType.ordinal() << 24);
+		}
+
+		@SuppressWarnings("unchecked")
+		static final class BuilderImpl<B extends Expression.Builder<B, E, R>, E extends ch.njol.skript.lang.Expression<R>, R>
+			extends SyntaxInfoImpl.BuilderImpl<B, E>
+			implements Expression.Builder<B, E, R> {
```

Comment:
You're making a mess with these type parameters, either leave some internal documentation or name them something nicer (we don't pay by the letter here)

Reply:
I have updated the documentation on the interfaces and inherited the Javadoc on the internal implementations - let me know if you would prefer to see more though.